fe/fe-core/src/main/java/com/starrocks/planner/BinlogScanNode.java

```diff
@@ -0,0 +1,201 @@
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * BinlogScanNode read data from binlog of SR table
+ */
+public class BinlogScanNode extends ScanNode {
+
+    private static final Logger LOG = LogManager.getLogger(BinlogScanNode.class);
+
+    private boolean isFinalized = false;
+    private OlapTable olapTable;
+    private List<Long> tabletIds;
+    private Set<Long> scanBackendIds;
+    private List<TScanRangeLocations> scanRanges;
+
+    public BinlogScanNode(PlanNodeId id, TupleDescriptor desc) {
+        super(id, desc, "BinlogScanNode");
+        this.tabletIds = new ArrayList<>();
+        this.scanBackendIds = new HashSet<>();
+        this.scanRanges = new ArrayList<>();
+        olapTable = (OlapTable) Preconditions.checkNotNull(desc.getTable());
+    }
+
+    @Override
+    protected void toThrift(TPlanNode msg) {
+        TBinlogScanNode binlogScan = new TBinlogScanNode();
+        binlogScan.setTuple_id(desc.getId().asInt());
+
+        TStreamScanNode streamScan = new TStreamScanNode();
+        streamScan.setBinlog_scan(binlogScan);
+        streamScan.setSource_type(StreamSourceType.BINLOG);
+
+        msg.setStream_scan_node(streamScan);
+        msg.setNode_type(TPlanNodeType.STREAM_SCAN_NODE);
+    }
+
+    protected TBinlogOffset getBinlogOffset(long tabletId) {
+        throw new NotImplementedException("TODO");
+    }
+
+    @Override
+    public void init(Analyzer analyzer) throws UserException {
+        super.init(analyzer);
+    }
+
+    @Override
+    public void computeStats(Analyzer analyzer) {
+        if (CollectionUtils.isNotEmpty(scanBackendIds)) {
+            numNodes = scanBackendIds.size();
+        }
+    }
+
+    @Override
+    public void finalizeStats(Analyzer analyzer) throws UserException {
+        if (isFinalized) {
+            return;
+        }
+        computeScanRanges();
+        computeStats(analyzer);
+        isFinalized = true;
+    }
+
+    @Override
+    public boolean canUsePipeLine() {
+        return true;
+    }
+
+    @Override
+    protected String getNodeExplainString(String prefix, TExplainLevel detailLevel) {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append(prefix).append("table: ").append(olapTable.getName());
+        sb.append(prefix).append("tabletList: %s").append(Joiner.on(",").join(tabletIds));
+        return sb.toString();
+    }
+
+    // TODO: support partition prune and bucket prune
+    public void computeScanRanges() throws UserException {
+        scanRanges = new ArrayList<>();
+        TabletInvertedIndex invertedIndex = GlobalStateMgr.getCurrentInvertedIndex();
+        long localBeId = -1;
+        long dbId = -1;
+        String dbName = null;
+        for (Partition partition : CollectionUtils.emptyIfNull(olapTable.getAllPartitions())) {
```

Comment:
Just a question, Can we ensure Binlog is consumed partition by partition, maybe it is confused if multi partitions are consumed at the same time?

Reply:
We need to define the consistency of multi-partitions and multi-tables in the end.